stored_file2.php

```diff
@@ -0,0 +1,314 @@
+    public $source;
+
+    protected $__moved;
+
+    public function actual_filename()
+    {
+        if ($this->__moved)
+            return static::resolve_file($this->filename);
+        return $this->source;
+    }
+
+    public function has_supported_extension($supported_exts = null)
+    {
+        if (!$this->extension)
+            $this->extension = static::parse_extension($this->filename);
+
+        if (!$this->extension)
+            return false;
+
+        if ($supported_exts && is_array($supported_exts) && count($supported_exts))
+        {
+            return in_array($this->extension,
+                $supported_exts);
+        }
+
+        return in_array($this->extension,
+            static::$__supported_exts);
+    }
+
+    public function exists()
+    {
+        return is_file($this->actual_filename());
+    }
+
+    public function size()
+    {
+        return filesize($this->actual_filename());
+    }
+
+    public function human_size()
+    {
+        $size = $this->size();
+
+        if ($size > static::SIZE_GIB)
+        {
+            $size = ($size / static::SIZE_GIB);
+            return sprintf('%.2f GiB', $size);
+        }
+
+        if ($size > static::SIZE_MIB)
+        {
+            $size = ($size / static::SIZE_MIB);
+            return sprintf('%.2f MiB', $size);
+        }
+
+        $size = ($size / static::SIZE_KIB);
+        return sprintf('%.2f KiB', $size);
+    }
+
+    public function detect_mime()
+    {
+        if ($this->mime) return $this->mime;
+        $this->mime = File_Util::detect_mime($this->actual_filename());
+        return $this->mime;
+    }
+
+    public function generate_filename($ext = null)
+    {
+        if (!$ext) $ext = $this->extension;
+
+        $hash = md5_file($this->source);
+        $dir1 = substr($hash, 0, 2);
+        $dir2 = substr($hash, 2, 2);
+        $name = substr($hash, 4, 28);
+
+        $this->filename_origin = build_path($dir1, $dir2, sf('%s.%s', $name, $ext));
+        $this->filename = build_path($dir1, $dir2, $name . uniqid( '.', true) . ".$ext");
+
+        return $this->filename;
+    }
+
+    public function read()
+    {
+        return file_get_contents($this->actual_filename());
+    }
+
+    public function move()
+    {
+        if ($this->__moved) return;
+        $origin = static::resolve_file($this->filename_origin, true);
+        $target = static::resolve_file($this->filename, true);
+        $dir = dirname($target);
+        if (!is_dir($dir)) mkdir($dir, 0755, true);
+        if (file_exists($origin)) {
+            link($origin, $target);
+        } else {
+            $this->filename = $this->filename_origin;
+            rename($this->source, $origin);
+        }
+        $this->__moved = true;
+    }
+
+    public function delete()
+    {
+        $filename = $this->actual_filename();
+
+        // Safe delete to avoid old filename patterns
+        if (substr_count($filename, '.') >= 2) return unlink($filename);
```

Comment:
This assumes that the base directory doesn't itself contain a `.`. 

`/home/www/files.system/files/ab/cd/eeeeeeeeeeeeeeeeeeee.jpg`

^ This would break because of the `.` in `files.system`.

Reply:
Additionally, I don't think you would ever be able to delete the original file, only the links to it.

Reply:
Ohhh you are right, I need to apply the check against $this->filename. It assigns the original file when it does not exists in public storage, so the app can delete that file if somebody decides to do it.